llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Enable curriculum learning by resuming with a different dataset.
+
+This callback is currently experimental. The API may change in the future.
+"""
+
+import logging
+from typing import Any, Dict, Iterable, Union
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+from streaming import StreamingDataset
+from torch.utils.data import DataLoader
+
+log = logging.getLogger(__name__)
+
+
+class CurriculumLearning(Callback):
+    """Starts an epoch with a different dataset when resuming from a checkpoint.
+
+    This callback is currently experimental. The API may change in the future.
+
+    Args:
+        dataset_index (int): The index of the dataset currently being used.
+        current_dataset_config (Dict): The configuration of the dataset currently
+            being used.
+        dataloader (Union[DataLoader, Iterable]): The training dataloader currently being used.
+    """
+
+    def __init__(self, dataset_index: int, dataloader: Union[DataLoader,
+                                                             Iterable],
+                 current_dataset_config: Dict):
+        self.dataset_index = dataset_index
+        self.saved_dataset_index = 0
+        self.all_dataset_configs = []
+        # The current dataset config is resolved and passed in train.py
+        self.current_dataset_config = current_dataset_config
+        self.new_dataset_setup = False
+        self.dataset_config_appended = False
+
+        # Must pass in dataset directly since it is not actually accessible at Event.INIT in
```

Comment:
Can we lazy init this so that this callback doesn't have to take in the dataloader? I understand what you need is not accessible on `INIT`, but we can we just defer to whenever the first time it is accessible is?

Reply:
Yeah, i'd prefer fixing the events in composer to support this new location that you need something than requiring the callback to take in the dataloader.

Reply:
word yeah makes sense. I wanted to enable this with as minimal code changes as possible, but I can add that in

Reply:
modified the callback, assuming `Event.BEFORE_LOAD` exists